src/kostentraeger/index.ts

```diff
@@ -0,0 +1,412 @@
+
+    ): KostentraegerFindResult | undefined {
+        /* only comb through those which are for the right Leistungserbringergruppe */
+        const leGruppe = leistungsartToLeistungserbringergruppeSchluessel(leistungsart)
+        const forLEGruppe = this.index.get(leGruppe)
+        if (!forLEGruppe) {
+            return
+        }
+
+        for (const [kassenart, institutionLists] of forLEGruppe) {
+            /* exclude institution lists that are not valid yet and if several valid ones are 
+               available, take only the most current one that is already valid */
+            const institutions = findMostCurrentValidInstitutionList(institutionLists, date)
+            if (!institutions) {
+                continue
+            }
+
+            // create map of IK -> Institution with only institutions that are valid at the given date
+            const institutionsIndex = getValidInstitutionsIndex(institutions, date)
+
+            const pflegekasse = institutionsIndex.get(pflegekasseIK)
+            if (!pflegekasse) {
+                continue
+            }
+
+            const kostentraeger = findKostentraeger(pflegekasse, institutionsIndex, leistungsart, location)
+            
+            if (paperDataType == 0) {
+
+                const datenannahmestellen = findDatenannahmestelle(kostentraeger, institutionsIndex, leistungsart, location)
+                if (!datenannahmestellen) {
+                    continue
+                }
+                
+                return {
+                    pflegekasse: pflegekasse,
+                    kostentraeger: kostentraeger,
+                    encryptTo: datenannahmestellen.encryptTo,
+                    sendTo: datenannahmestellen.sendTo
+                }
+
+            } else {
+
+                const sendTo = findPapierannahmestelle(kostentraeger, institutionsIndex, paperDataType, leistungsart, location)
+                if (!sendTo) {
+                    return
+                }
+    
+                return {
+                    pflegekasse: pflegekasse,
+                    kostentraeger: kostentraeger,
+                    sendTo: sendTo
+                }
+
+            }
+        }
+    }
+}
+
+/** Finds the index of the entry in the given lists that has the most current valid date that is
+ *  at the same time already valid at the given date.
+ */
+function findMostCurrentValidInstitutionList(lists: InstitutionListWithValidityStartDate[], date: Date): Institution[] | undefined {
+    let result: Institution[] | undefined = undefined
+    let mostCurrentValidityStartDate: Date | undefined
+    for (let i = 0; i < lists.length; i++) {
+        const list = lists[i]
+        if (list.validityStartDate < date) {
+            if (!mostCurrentValidityStartDate || list.validityStartDate > mostCurrentValidityStartDate) {
+                result = list.institutions
+                mostCurrentValidityStartDate = list.validityStartDate
+            }
+        }
+    }
```

Comment:
I'd suggest two minor simplifications:

```suggestion
    let mostCurrentValidityStartDate = new Date(0) // 1970
    lists.forEach(list => {
        if (list.validityStartDate < date && list.validityStartDate > mostCurrentValidityStartDate) {
            result = list.institutions
            mostCurrentValidityStartDate = list.validityStartDate
        }
    })
```

Reply:
Ah right. The function in its first version returned an `indexOf`, that's why there is still the fori-loop.